### DISCUSSION TITLE
SQLAlchemy subquery not mapped to model

### DISCUSSION THREAD

**Discussion by woutrbe** [sqlalchemy/sqlalchemy#10565] [Usage Questions] [answered]
I have a `User` and a `UserConfig` model, a `User` can have many `UserConfig`, but only the last created `UserConfig` should be used. The rest are stored for historic references.

So my models look like this:
    
    class User(Base):
        __tablename__ = "user"

        id = mapped_column(String, primary_key=True)
        type = mapped_column(String, nullable=False)
        description = mapped_column(String, nullable=True)

        config = relationship("UserConfig")

        created_at = mapped_column(DateTime, server_default=func.now())

    class UserConfig(Base):
        __tablename__ = "user_config"

        id = mapped_column(Integer, primary_key=True, autoincrement=True)

        user_id = mapped_column(String, ForeignKey("user.id", ondelete="CASCADE"), name="user", nullable=False)
        user = relationship("User", back_populates="config")

        weight = mapped_column(Integer, nullable=False)
        meta = mapped_column(JSON, nullable=False)

        created_at = mapped_column(DateTime, server_default=func.now())

I now want to write a query which selects the `User`, and the latest `UserConfig`. This is what I currently have:

```
user_config_stm = session.query(UserConfig).distinct(UserConfig.user_id).order_by(UserConfig.user_id).order_by(desc(UserConfig.created_at)).subquery()
result = session.query(User, user_config_stm).where(User.id == id).join(user_config_stm, user_config_stm.c.user == id).first()
```

This however returns the following tuple: `(<apps.core.sql.models.user.User object at 0x11c904160>, 14, 'xxx', 12, '""', datetime.datetime(2023, 10, 29, 22, 52, 28, 860854))`. 

I can't figure out why the results of the subquery are not mapped back to a `UserConfig` model. Ideally what I want is a tuple of `(User, UserConfig)`

**Comment by zzzeek** [comment, marked as answer]
so the kind of operation you're doing is more reasonably documented if you [use new style queries](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#orm-query-unified-with-core-select), though the same caveats apply, which is that subqueries are automatically assumed to be "selectables" (things with columns) and not "entities" (things that represent a single class).  you can see this because you are using the `c` attribute at `user_config_stm.c.user == id`.

to have `user_config_stm` yield `UserConfig` objects you want to use `aliased()` and use that throughout the query.

```
user_config_stm = session.query(UserConfig).distinct(UserConfig.user_id).order_by(UserConfig.user_id).order_by(desc(UserConfig.created_at)).subquery()
user_config_alias = aliased(UserConfig, user_config_stm)
result = session.query(User, user_config_alias).where(User.id == id).join(user_config_alias, user_config_alias.user == id).first()
```

background on this which has been reworked to be in terms of new style queries, though the idea is basically the same, is at https://docs.sqlalchemy.org/en/20/orm/queryguide/select.html#selecting-entities-from-subqueries

**Reply by woutrbe** [reply]
Thank you very much for that, this finally solved the issues I was facing. I'll also look into upgrading all my queries to the new style queries, although I find them slightly more confusing than the old version.

**Reply by woutrbe** [reply]
I've re-written it into the new style queries, and ended up with the following, which is working as expected.

```
user_config_stm = select(UserConfig).distinct(UserConfig.label_id).order_by(UserConfig.user_id, desc(UserConfig.created_at)).subquery()
user_config_alias = aliased(UserConfig, user_config_stm)
(user, user_config) = session.execute(select(User, user_config_alias).where(User.id == id).join(user_config_alias, user_config_alias.user_id == id)).first()
```
